### PR TITLE
Auto shutdown GCP instance

### PR DESCRIPTION
**Pull request by tapih**
https://github.com/cybozu-go/neco/issues/702

### CODE REVIEW

File: gcp/compute.go [hunk 1]
```diff
@@ -6,22 +6,39 @@ import (
...
+
+at {{ .ShutdownAt }} -f $STARTUP_PATH
+`))
+
 const (
 	retryCount   = 300
 	imageLicense = "https://www.googleapis.com/compute/v1/projects/vm-options/global/licenses/enable-vmx"
 	// MetadataKeyExtended is key for extended time in metadata
 	MetadataKeyExtended = "extended"
+	startUpScriptPath   = "/tmp/gcp-instance-startup.sh"
```
Comment (ymmt2005) [inline gcp/compute.go]: Why don't you use [`ioutil.Tempfile`](https://golang.org/pkg/io/ioutil/#TempFile)?

File: gcp/compute.go [hunk 2]
```diff
@@ -88,18 +105,54 @@ func (cc *ComputeClient) CreateVMXEnabledInstance(ctx context.Context) error {
...
 
+func convertLocalTimeToUTC(timezone, shutdownAt string) (string, error) {
+	loc, err := time.LoadLocation(timezone)
+	if err != nil {
+		return "", err
+	}
+	now := time.Now().In(loc)
+	localTime := fmt.Sprintf("%d-%02d-%02d "+shutdownAt+":00",
+		now.Year(), now.Month(), now.Day())
+	t, err := time.ParseInLocation(timeFormat, localTime, loc)
```
Comment (ymmt2005) [inline gcp/compute.go]: Ain't it equal to the following?

```go
t, err := time.ParseInLocation("15:04", shutdownAt, loc)
```

Comment (tapih) [reply gcp/compute.go]: As a result of an experiment, I found that time format also needs a year and date to recognize timezone correctly.
`time.ParseInLocation("15:04", shutdownAt, loc)` gets the output like below:

```
0000-01-01 21:00:00 +0918 LMT
```

File: gcp/compute.go [hunk 3]
```diff
@@ -88,18 +104,62 @@ func (cc *ComputeClient) CreateVMXEnabledInstance(ctx context.Context) error {
...
+		ShutdownAt: shotdownAt,
+	})
+	if err != nil {
+		return err
+	}
+	tmpfile, err := ioutil.TempFile("/tmp", "gcp-start-up-script-*.sh")
+	if err != nil {
+		return err
+	}
+	log.Info("start up script for "+tmpfile.Name(), map[string]interface{}{})
```
Comment (ymmt2005) [inline gcp/compute.go]: should delete the temporary file as follows:

```go
defer func() {
    tmpfile.Close()
    os.Remove(tmpfile.Name())
}()
```

File: gcp/compute.go [hunk 4]
```diff
@@ -88,18 +104,62 @@ func (cc *ComputeClient) CreateVMXEnabledInstance(ctx context.Context) error {
...
+	_, err = tmpfile.Write(buf.Bytes())
+	if err != nil {
+		return err
```
Comment (ymmt2005) [inline gcp/compute.go]: This leaks an open file descriptor!
Remember that open files should always be closed in a `defer` clause.

**Pull request merged**